velox/dwio/parquet/reader/TimestampColumnReader.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/dwio/parquet/reader/IntegerColumnReader.h"
+#include "velox/dwio/parquet/reader/ParquetColumnReader.h"
+
+namespace facebook::velox::parquet {
+
+class TimestampColumnReader : public IntegerColumnReader {
+ public:
```

Comment:
Can we create a `struct int96 {int64_t val[3];};` in this class and use that instead of allocation `int128_t`?

Reply:
Just `char[12]` (or a local `using int96_t = char[12]`) would be good I think

Reply:
Hi @Yuhta @majetideepak, I tried above two ways, but I have to say a lot of compilation errors were encountered. 

For `char[12]` way, below error was found during compilation.
```shell
../.././velox/dwio/common/SelectiveColumnReaderInternal.h:55:39:   required from ‘void facebook::velox::dwio::common::SelectiveColumnReader::ensureValuesCapacity(facebook::velox::vector_size_t) [with T = char [12]; facebook::velox::vector_size_t = int]’
../.././velox/dwio/common/SelectiveColumnReaderInternal.h:100:3:   required from ‘void facebook::velox::dwio::common::SelectiveColumnReader::prepareRead(facebook::velox::vector_size_t, facebook::velox::dwio::common::RowSet, const uint64_t*) [with T = char [12]; facebook::velox::vector_size_t = int; facebook::velox::dwio::common::RowSet = folly::Range<const int*>; uint64_t = long unsigned int]’
../.././velox/dwio/parquet/reader/TimestampColumnReader.h:41:47:   required from here
/usr/include/c++/9/optional:952:2: error: function returning an array
  952 |  value_or(_Up&& __u) const&
      |  ^~~~~~~~
/usr/include/c++/9/optional:963:2: error: function returning an array
  963 |  value_or(_Up&& __u) &&
      |  ^~~~~~~~
In file included from ../../velox/dwio/parquet/reader/ParquetColumnReader.cpp:22:
../.././velox/dwio/common/SelectiveColumnReaderInternal.h: In instantiation of ‘void facebook::velox::dwio::common::SelectiveColumnReader::ensureValuesCapacity(facebook::velox::vector_size_t) [with T = char [12]; facebook::velox::vector_size_t = int]’:
../.././velox/dwio/common/SelectiveColumnReaderInternal.h:100:3:   required from ‘void facebook::velox::dwio::common::SelectiveColumnReader::prepareRead(facebook::velox::vector_size_t, facebook::velox::dwio::common::RowSet, const uint64_t*) [with T = char [12]; facebook::velox::vector_size_t = int; facebook::velox::dwio::common::RowSet = folly::Range<const int*>; uint64_t = long unsigned int]’
../.././velox/dwio/parquet/reader/TimestampColumnReader.h:41:47:   required from here
../.././velox/dwio/common/SelectiveColumnReaderInternal.h:55:39: note:   when instantiating default argument for call to ‘static facebook::velox::BufferPtr facebook::velox::AlignedBuffer::allocate(size_t, facebook::velox::memory::MemoryPool*, const std::optional<_Tp>&) [with T = char [12]; facebook::velox::BufferPtr = boost::intrusive_ptr<facebook::velox::Buffer>; size_t = long unsigned int]’
   55 |   values_ = AlignedBuffer::allocate<T>(
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~^
   56 |       numRows + (simd::kPadding / sizeof(T)), &memoryPool_);
      |       ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
For `struct int96 {int val[3];};` way,  I solved some compilation errors and ended up with an error of missing conversion from `int64_t` to `int96_t` at [RLEv1.h](https://github.com/facebookincubator/velox/blob/main/velox/dwio/dwrf/common/RLEv1.h#L310). Because I create a columnar visitor in [SelectiveIntegerColumnReader.h](https://github.com/facebookincubator/velox/blob/main/velox/dwio/common/SelectiveIntegerColumnReader.h#L97-L102) with type int96_t (the self-defined struct), while `value` is of int64_t type. 

```c++
    case 12:
      reinterpret_cast<Reader*>(this)->Reader::readWithVisitor(
          rows,
          ColumnVisitor<int96_t, TFilter, ExtractValues, isDense>(
              *reinterpret_cast<TFilter*>(filter), this, rows, extractValues));
      break;
```

Looks more work is needed here to support a self-defined int96 type, and some template functions for integer need to be changed.
Do you have any suggestion here?

Reply:
Can we consider using `int128_t` as an alternative first? Because it helps simply the implementation.

Reply:
I don't think using `int128_t` will work here, the `valueSize_` is different and you will end up reading different part of data and even read out of bound.

Maybe check https://github.com/facebookincubator/velox/pull/2822 as a reference for what need to be done.

Reply:
@Yuhta Thank you. I see. The main problem here is `__int128` is a built-in type in GCC while `int96` is not there, so to use `int96_t` here more work. For example, the implicit conversions between `int64_t` and `int96_t`, and some operators like `*` and `+=` etc. are lacking. Therefore, a lot of compiling errors occur in the code path of `IntegerColumnReader`.
Do we intend to implement a robust `int96` type, or try other solution if there is?

Reply:
The actual underlying structure is not really a generic `int96_t` but this:
```c++
struct __attribute__((__packed__)) Int96Timestamp {
  int32_t days;
  uint64_t nanos;
};
```

I would recommend you to make this type to work with `IntegerColumnReader`, that would be the easiest way I guess.

Reply:
I will continue to work on that. Convert this PR as draft for now, and will ping you for review when it's ready. Thanks.